crates/sui-core/src/state_accumulator.rs

```diff
@@ -0,0 +1,588 @@
+            Ok(())
+        } else {
+            self.state_hash_by_checkpoint
+                .insert(&checkpoint, &state_hash)
+        }
+    }
+
+    pub fn get_highest_hashed_checkpoint_seq_number(
+        &self,
+    ) -> Result<Option<CheckpointSequenceNumber>, TypedStoreError> {
+        if let Some(highest_hashed) = self.watermarks.get(&CheckpointWatermark::HighestHashed)? {
+            Ok(Some(highest_hashed))
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn get_highest_accumulated_checkpoint_seq_number(
+        &self,
+    ) -> Result<Option<CheckpointSequenceNumber>, TypedStoreError> {
+        if let Some(highest_accumulated) = self
+            .watermarks
+            .get(&CheckpointWatermark::HighestAccumulated)?
+        {
+            Ok(Some(highest_accumulated))
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn get_highest_checkpoint_ready_to_hash_seq_number(
+        &self,
+    ) -> Result<Option<CheckpointSequenceNumber>, TypedStoreError> {
+        if let Some(highest_ready) = self
+            .watermarks
+            .get(&CheckpointWatermark::HighestReadyToHash)?
+        {
+            Ok(Some(highest_ready))
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn update_highest_hashed_checkpoint_seq_number(
+        &self,
+        checkpoint_seq_num: &CheckpointSequenceNumber,
+    ) -> Result<(), TypedStoreError> {
+        self.watermarks
+            .insert(&CheckpointWatermark::HighestHashed, checkpoint_seq_num)
+    }
+
+    pub fn update_highest_accumulated_checkpoint_seq_number(
+        &self,
+        checkpoint_seq_num: &CheckpointSequenceNumber,
+    ) -> Result<(), TypedStoreError> {
+        self.watermarks
+            .insert(&CheckpointWatermark::HighestAccumulated, checkpoint_seq_num)
+    }
+
+    pub fn update_highest_checkpoint_ready_to_hash_seq_number(
+        &self,
+        checkpoint_seq_num: &CheckpointSequenceNumber,
+    ) -> Result<(), TypedStoreError> {
+        self.watermarks
+            .insert(&CheckpointWatermark::HighestReadyToHash, checkpoint_seq_num)
+    }
+}
+
+/// Thread safe clone-able handle for enqueueing items to Accumulator
+#[derive(Clone)]
+pub struct EnqueueHandle {
+    sender: mpsc::Sender<State>,
+    store: Arc<StateAccumulatorStore>,
+}
+
+impl EnqueueHandle {
+    /// Enqueue checkpoint state for accumulation. A checkpoint shall
+    /// be processed (accumulated both individually and towards
+    /// the root state hash) exactly once
+    ///
+    /// NOTE: When called from executor, this should be called before we increment
+    /// highest executed checkpoint watermark, otherwise we may inefficiently read effects
+    /// from the effects store rather than using effects passed in memory throuhg enqueue.
+    /// This may happen due to a race between scheduling new checkpoints for hashing and
+    /// checking the queue.
+    pub async fn enqueue(&self, state: State) -> SuiResult {
```

Comment:
Whoever is calling this knows what checkpoint they are enqueuing - we should pass that in rather than relying on watermarks to determine which checkpoint is being enqueued. This may allow you to remove one of the watermarks as well, not sure about that though.

Reply:
Not sure I'm following. `State` does contain the checkpoint sequence number, so it is already being passed in. And the watermark is for internal tracking so that we don't overschedule (among other things)